### ISSUE TITLE
Java System.out not printed for non-main threads

### ISSUE THREAD

**Issue by apohllo** [twosigma/beakerx#6617]
I've run the following code for a Java kernel
```java
import static java.lang.System.out;
Thread t = new Thread(() -> out.println("Some output"));
t.start();
t.join();
```
But there is no output produced.

Please let me know if there is any workaround (even messy) that could be used in this case, to get any output from a thread.

**Comment by scottdraves** [comment]
Thank you, you are right, output from aux threads goes to the console, not into the notebook.
Perhaps that's enough for your use case?  I would have to know more about it.
Perhaps you can install your own output that writes to a file for example:
https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#setOut-java.io.PrintStream-

This wasn't by design, but might be hard to fix.  Beaker Notebook was able to catch this output and show it in the output tray, which collected any output not associated with a specific cell.
But Jupyter doesn't have a tray....  Maybe we could make one.

**Comment by apohllo** [comment]
Well, I have Java lectures and I use jupyter to run the code during the presentation. The output of the other threads is crucial for explaining the students how the threads interact with each other. Thus writing to a file is not an option.

**Comment by apohllo** [comment]
Maybe a solution would be to add some "magic" to:
1. Catch output from any thread as the output of the running cell.
2. Wait until all threads are finished before finishing the running of the current cell.

In fact - the other Java kernels that were available (but no longer are) implemented 2 by default. I understand that this is not a good default for beakerx, but a magic switch would do the work. If this sounds sensible to you, please let me know where I should look in order to implement such a strategy.

**Comment by altavir** [comment]
I think it is rather simple to fix. You just need to replace standard system out with something like this (kotlin notation to minimize code):
```kotlin
class BeakerPrintStream(val out: PrintStream): PrintStream{
   val executor: ExecutorService = Executors.newSingleThreadExecutor()

   fun print(...){
      executor.submit{
           out.print(...)
      }
   }
}
```

You can also check if the thread is main thread, and, say display other text with another color or something like that if it is not.

Generally, all ui actions should be run on specific ui thread and print to system out in case of beaker is definitely ui action.

**Comment by jmsdnns** [comment]
Maybe yall would get some ideas from this project, which implemented similar thinking for python kernels: https://github.com/micahscopes/nbmultitask/.

In this Python case, they use a Queue from the multiprocessing module that each thread writes to. _It's actually a process because Python threading is bad, but I'm going to continue saying "thread" for simplicity_. The Queue itself coordinates passing data from the threads to the main thread and the main thread then writes it all out to the Notebook.

It is has similarities to @altavir's suggestion of replacing stdout. Each Python thread replaces stdout and stderr with something that writes to the Queue instead, so the threads barely notice anything changed.

It looks as though ipywidgets can do the rest just fine by nature of the multiplexing happening kernel-side.

Hopefully this help inform the designs yall are thinking about.

**Comment by scottdraves** [comment]
Thanks that's cool.  Also the standard output widget (see http://ipywidgets.readthedocs.io/en/stable/examples/Widget%20List.html) has a similar idea.
We have Java APIs for most of the standard widgets, but not this one.  That could be fixed though.

**Comment by scottdraves** [comment]
eg: first create the output widget, redirect to it, and then display it:
```
import com.twosigma.beakerx.widgets.Output
out = new Output()
beakerx.setOutput(out)
out
```
then start a thread
```
import static java.lang.System.out;
Thread t = new Thread(() -> Thread.sleep(10000); out.println("Some output"));
t.start();
```
after 10s, "Some output" appears in the 1st cell.  `beakerx.setOutput(null)` should return output going to current cell if there is one.

**Comment by apohllo** [comment]
Wow, thank you vary much indeed! What is the easiest way to test this feature on my machine? I'm using anaconda.

**Comment by scottdraves** [comment]
np, thank you for alerting us to this problem.  you can see what we did at the bottom of this notebook: https://github.com/twosigma/beakerx/blob/master/test/notebooks/java/JavaTest.ipynb

right now you would have to build from master following the instructions in the readme.
shouldn't be hard.

if you wait about a week it should be released in 0.13.0.
